File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
from collections import defaultdict 

def solution(gems):
    answer = [100001, 100001]
    sequence_len = len(gems) + 10
    
    # end_point를 늘리면서 보석을 담음
    required_kind_cnt = len(set(gems))
    getting_gem_dict = defaultdict(int)
    
    cur_cnt = 0 
    gems_len = len(gems)
    end_point = 0 
    
    for start_point in range(gems_len):
#         cur_gem = gems[start_point]
        
#         if( 0 == getting_gem_dict[cur_gem]):
#             cur_cnt += 1
            
#         getting_gem_dict[cur_gem] += 1
        
        while cur_cnt < required_kind_cnt and end_point < gems_len:
            cur_gem = gems[end_point]
        
            if( 0 == getting_gem_dict[cur_gem]):
                cur_cnt += 1
            
            getting_gem_dict[cur_gem] += 1
            
            end_point += 1
        
        cur_sequence_len = end_point - start_point
        
        if((cur_cnt == required_kind_cnt)):
            if(sequence_len > cur_sequence_len):
                sequence_len = cur_sequence_len
                answer = [start_point, end_point]
            elif(answer[0] > start_point):
                answer = [start_point, end_point]
            else:
                pass
        cur_gem = gems[start_point]
        
        if( 1 == getting_gem_dict[cur_gem]):
            cur_cnt -= 1        
        getting_gem_dict[cur_gem] -= 1 

    return [answer[0] + 1, answer[1] ]

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> gems) {
    
    vector<int> answer;
    map <string, int> pocketMap;
    
    for(vector<string>::iterator it = gems.begin(); it != gems.end(); ++it){
        pocketMap[*it] = 0;
    }
    
    int gemCnt = pocketMap.size();
    int startIdx = 0; int endIdx = 0;
    
    int cpyGemCnt = gemCnt;
    int partLen = 100002;
    
    while((startIdx <= endIdx) && ( endIdx < gems.size())){

        const string& curGem = gems[endIdx];
        int & curCnt = pocketMap[curGem];

        // 현재 보석이 주머니에 없음
        // 새로운 종류의 보석
        if(curCnt == 0){
            --cpyGemCnt;
        }
        
        curCnt += 1;

        // 모든 종류의 보석을 수집했으면 startIdx를 옮겨줌
        while(cpyGemCnt <= 0){
            int curPartLen = endIdx - startIdx;
            
            // 현재 구간이 더 짧으면 갱신
            if(partLen > curPartLen){
                partLen = curPartLen;
                answer = {startIdx + 1, endIdx + 1};
            }
            
            // startIdx 위치의 보석을 주머니에서 빼자.
            const string & prevGem = gems[startIdx];
            int & prevCnt = pocketMap[prevGem];
            prevCnt -= 1;
            // prevCnt가 0이 되면 주머니에 해당 보석이 없음
            cpyGemCnt += (prevCnt == 0);
            // startIdx 옮기기 
            ++startIdx;
        }
        endIdx += 1;
    }
    
    
    // gems에 있는 unique한 보석 이름을 찾아야 함 
    return answer;
}